### ISSUE TITLE
complete_inserts gets turned on in the during the dump // Generated Column is mentioned in the VALUES but not on the fields -> Column count doesn't match value count

### ISSUE THREAD

**Issue by amenk** [Smile-SA/gdpr-dump#78]
### Preconditions
GdprDump Version: 3.1.0
PHP Version: 8.1.
Database Version: MariaDB 10.3.34

### Steps to reproduce

1. Export a Shopware database with a blog plugin installed
2. Try to import

### Expected result

1. Import successfull

### Actual result
1. ERROR 1136 (21S01) at line 9250: Column count doesn't match value count at row 1

### This is excerpt of the dump:

```
CREATE TABLE `h1webblog_blog` (
  `id` binary(16) NOT NULL,
  `media_id` binary(16) DEFAULT NULL,
  `created_at` datetime(3) NOT NULL,
  `updated_at` datetime(3) DEFAULT NULL,
  `option_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `property_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `properties` binary(16) DEFAULT NULL,
  `category_tree` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `categories` binary(16) DEFAULT NULL,
  `tags` binary(16) DEFAULT NULL,
  `tag_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `published_at` datetime DEFAULT current_timestamp(),
  `custom_fields` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `cms_page_version_id` binary(16) DEFAULT NULL,
  `cms_page_id` binary(16) DEFAULT NULL,
  `visibilities` binary(16) GENERATED ALWAYS AS (`id`) VIRTUAL,
  PRIMARY KEY (`id`),
  KEY `fk.h1webblog_blog.cms_page_id` (`cms_page_id`),
  KEY `published_at_index` (`published_at`),
  CONSTRAINT `fk.h1webblog_blog.cms_page_id` FOREIGN KEY (`cms_page_id`) REFERENCES `cms_page` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `h1webblog_blog`
--

LOCK TABLES `h1webblog_blog` WRITE;
/*!40000 ALTER TABLE `h1webblog_blog` DISABLE KEYS */;
SET autocommit=0;
INSERT INTO `h1webblog_blog` (`id`, `media_id`, `created_at`, `updated_at`, `option_ids`, `property_ids`, `properties`, `category_tree`, `categories`, `tags`, `tag_ids`, `published_at`, `custom_fields`, `cms_page_version_id`, `cms_page_id`) VALUES
(0x522438891C444F69BA39539FA154351D,0x8E2D23F8F09449AF8022B9C136FEB9C1,'2022-09-09 12:33:32.649','2022-11-09 15:33:48.141',NULL,'[\"22b5567892aa41bdb8eefcb79c59b221\"]',NULL,'[\"d6715fe94e8e401eb04a2c0e6078d5ff\",\"c664685f17464b3fa55af46af7d0f308\",\"99d8198817e84265976bcc541578dfa8\"]',NULL,NULL,'[\"0dfb3651cc24417788e9c76100ad25b7\",\"6bf496fc9e364946b3f589c0d46a538e\",\"bdd86e333c7449669e632a4a1075a269\"]','2029-09-13 12:00:00',NULL,0x0FA91CE3E96A4BC2BE4BD9CE752C3425,0x42190FE518E14BB48145102C111DEA33,0x522438891C444F69BA39539FA154351D),

```

The generated column is not mentioned in the field list, but the value seems to be mentioned on the values list.

**Comment by amenk** [comment]
Maybe it happens around here? https://github.com/ifsnop/mysqldump-php/commit/8d063449981e3f7bd28a3a083842e21a24d1f4af#diff-2bdc9acc0b460325abf5e827d6c50a098f55c2d555a5aa5fb847f545b230d3e5R927

**Comment by amenk** [comment]
What is confusing for us: In older dumps we did not have the field list in the INSERTs - the question is why this appeared now, we cannot remember changing anything in the config.

**Comment by guvra** [comment]
> What is confusing for us: In older dumps we did not have the field list in the INSERTs - the question is why this appeared now, we cannot remember changing anything in the config.

The last release of mysqldump-php was in April 2020 so that's surprising.

We might need to switch to a maintained fork of mysqldump-php, maybe https://github.com/druidfi/mysqldump-php

**Comment by amenk** [comment]
I am still confused. I tried all old 2.x versions but cannot reproduce our old dumps we did a month a ago. In those dumps the insert looks like this:

```
CREATE TABLE `h1webblog_blog` (
  `id` binary(16) NOT NULL,
  `media_id` binary(16) DEFAULT NULL,
  `created_at` datetime(3) NOT NULL,
  `updated_at` datetime(3) DEFAULT NULL,
  `option_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `property_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `properties` binary(16) DEFAULT NULL,
  `category_tree` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `categories` binary(16) DEFAULT NULL,
  `tags` binary(16) DEFAULT NULL,
  `tag_ids` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `published_at` datetime DEFAULT current_timestamp(),
  `custom_fields` longtext CHARACTER SET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL,
  `cms_page_version_id` binary(16) DEFAULT NULL,
  `cms_page_id` binary(16) DEFAULT NULL,
  `visibilities` binary(16) GENERATED ALWAYS AS (`id`) VIRTUAL,
  PRIMARY KEY (`id`),
  KEY `fk.h1webblog_blog.cms_page_id` (`cms_page_id`),
  KEY `published_at_index` (`published_at`),
  CONSTRAINT `fk.h1webblog_blog.cms_page_id` FOREIGN KEY (`cms_page_id`) REFERENCES `cms_page` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `h1webblog_blog`
--

LOCK TABLES `h1webblog_blog` WRITE;
/*!40000 ALTER TABLE `h1webblog_blog` DISABLE KEYS */;
SET autocommit=0;
INSERT INTO `h1webblog_blog` VALUES
(0x522438891C444F69BA39539FA154351D,0x8E2D23F8F09449AF8022B9C136FEB9C1,'2022-09-09 12:33:32.649','2022-10-04 10:49:41.910',NULL,'[\"22b5567892aa41bdb8eefcb79c59b221\"]',NULL,'[\"d6715fe94e8e401eb04a2c0e6078d5ff\",\"c664685f17464b3fa55af46af7d0f308\",\"99d8198817e84265976bcc541578dfa8\"]',NULL,NULL,'[\"0dfb3651cc24417788e9c76100ad25b7\",\"6bf496fc9e364946b3f589c0d46a538e\",\"bdd86e333c7449669e632a4a1075a269\"]','2022-09-02 12:00:00',NULL,0x0FA91CE3E96A4BC2BE4BD9CE752C3425,0x42190FE518E14BB48145102C111DEA33,0x522438891C444F69BA39539FA154351D),
```

so the fields at not mentioned and the query executes.

I cannot reproduce anymore how that happend because we don't have a backup of the `gdpr-dump.phar` used at this time.

Anyway, to move forward we should fix the dumps of generated columns. Do you know if that is fixed in the fork you mentioned? Shall we just switch to it?

**Comment by guvra** [comment]
> I cannot reproduce anymore how that happend because we don't have a backup of the gdpr-dump.phar used at this time.

Maybe you changed something in the [dump settings](https://github.com/Smile-SA/gdpr-dump/blob/master/docs/01-configuration.md#dump-settings) of your config file (e.g. `complete-insert`).

> Anyway, to move forward we should fix the dumps of generated columns. Do you know if that is fixed in the fork you mentioned? Shall we just switch to it?

Probably not, very few php frameworks use this kind of feature.

**Comment by amenk** [comment]
> > I cannot reproduce anymore how that happend because we don't have a backup of the gdpr-dump.phar used at this time.
> 
> Maybe you changed something in the [dump settings](https://github.com/Smile-SA/gdpr-dump/blob/master/docs/01-configuration.md#dump-settings) of your config file (e.g. `complete_insert`).

we have our config in git and it was not changed since months.
But if I set `complete_insert: false` explicitly it produces such a dump without field names.
Interestingly false is the default, so it should always behave like that.

Can this default be influenced by anythings else?

Anyways, I would have a workaround now. Thank you.

> > Anyway, to move forward we should fix the dumps of generated columns. Do you know if that is fixed in the fork you mentioned? Shall we just switch to it?
>
> Probably not, very few php frameworks use this kind of feature.

Shopware 6 does :-)

**Comment by amenk** [comment]
Concerning the first point (sorry, I cannot let this go, it's really strange)

I looks like, without the `complete_insert` flag set to anything or to false, half of the database is dumped without field names, then suddenly it starts to dump with field names.

My current dump looks like this (with the same config as a month ago)

```
$ grep "INSERT INTO" sql/master.sql 
INSERT INTO `h1webblog_blog_search_keyword` VALUES
INSERT INTO `property_group_option` VALUES
INSERT INTO `number_range_sales_channel` VALUES
INSERT INTO `neti_export_on_event_action` VALUES
INSERT INTO `flow_sequence` VALUES
INSERT INTO `media_translation` VALUES
INSERT INTO `category_translation` VALUES
INSERT INTO `product_translation` VALUES
INSERT INTO `h1webblog_category` VALUES
INSERT INTO `h1webblog_tag` VALUES
INSERT INTO `sales_channel_domain` VALUES
INSERT INTO `document_base_config` VALUES
INSERT INTO `cms_slot` VALUES
INSERT INTO `product_search_keyword` VALUES
INSERT INTO `state_machine_state_translation` VALUES
INSERT INTO `swag_migration_run` VALUES
INSERT INTO `import_export_profile` VALUES
INSERT INTO `import_export_profile_translation` VALUES
INSERT INTO `product_feature_set_translation` VALUES
INSERT INTO `state_machine_translation` VALUES
INSERT INTO `product` (`id`, `version_id`, `auto_increment`, `product_number`, `active`, `parent_id`, `parent_version_id`, `tax_id`, `product_manufacturer_id`, `product_manufacturer_version_id`, `delivery_time_id`, `deliveryTime`, `product_media_id`, `product_media_version_id`, `cms_page_id`, `cms_page_version_id`, `unit_id`, `product_feature_set_id`, `category_tree`, `category_ids`, `stream_ids`, `option_ids`, `property_ids`, `tax`, `manufacturer`, `cover`, `unit`, `media`, `prices`, `visibilities`, `properties`, `categories`, `translations`, `price`, `manufacturer_number`, `ean`, `sales`, `stock`, `available_stock`, `available`, `restock_time`, `is_closeout`, `purchase_steps`, `max_purchase`, `min_purchase`, `purchase_unit`, `reference_unit`, `shipping_free`, `purchase_prices`, `mark_as_topseller`, `weight`, `width`, `height`, `length`, `release_date`, `whitelist_ids`, `blacklist_ids`, `tag_ids`, `tags`, `variant_restrictions`, `configurator_group_config`, `main_variant_id`, `created_at`, `updated_at`, `rating_average`, `display_group`, `child_count`, `crossSellings`, `featureSet`, `customFieldSets`, `custom_field_set_selection_active`, `canonical_product_id`, `canonicalProduct`, `cmsPage`, `cheapest_price`, `cheapest_price_accessor`, `display_parent`) VALUES
INSERT INTO `cms_block` (`id`, `version_id`, `cms_section_id`, `cms_section_version_id`, `position`, `section_position`, `type`, `name`, `locked`, `margin_top`, `margin_bottom`, `margin_left`, `margin_right`, `background_color`, `background_media_id`, `background_media_mode`, `css_class`, `custom_fields`, `created_at`, `updated_at`) VALUES
INSERT INTO `media_folder` (`id`, `parent_id`, `default_folder_id`, `name`, `child_count`, `path`, `media_folder_configuration_id`, `use_parent_configuration`, `custom_fields`, `created_at`, `updated_at`) VALUES
INSERT INTO `novalnet_transaction_details` (`id`, `tid`, `payment_type`, `amount`, `currency`, `paid_amount`, `refunded_amount`,
.....
```


The old dump looks like this:

```
# grep "INSERT INTO" sql/master.sql 
INSERT INTO `h1webblog_blog_search_keyword` VALUES
INSERT INTO `property_group_option` VALUES
INSERT INTO `number_range_sales_channel` VALUES
INSERT INTO `neti_export_on_event_action` VALUES
INSERT INTO `flow_sequence` VALUES
INSERT INTO `media_translation` VALUES
INSERT INTO `category_translation` VALUES
INSERT INTO `product_translation` VALUES
INSERT INTO `promotion_discount` VALUES
INSERT INTO `h1webblog_category` VALUES
INSERT INTO `h1webblog_tag` VALUES
INSERT INTO `sales_channel_domain` VALUES
INSERT INTO `document_base_config` VALUES
INSERT INTO `cms_slot` VALUES
INSERT INTO `product_search_keyword` VALUES
INSERT INTO `state_machine_state_translation` VALUES
INSERT INTO `swag_migration_run` VALUES
INSERT INTO `import_export_profile` VALUES
INSERT INTO `import_export_profile_translation` VALUES
INSERT INTO `product_feature_set_translation` VALUES
INSERT INTO `state_machine_translation` VALUES
INSERT INTO `product` VALUES
INSERT INTO `cms_block` VALUES
INSERT INTO `media_folder` VALUES
INSERT INTO `novalnet_transaction_details` VALUES
INSERT INTO `media_thumbnail_size` VALUES
INSERT INTO `product_cross_selling` VALUES
INSERT INTO `sales_channel_api_context` VALUES
INSERT INTO `app_config` VALUES
INSERT INTO `media_folder_configuration_media_thumbnail_size` VALUES
INSERT INTO `h1webblog_blog` VALUES
INSERT INTO `payment_token` VALUES
INSERT INTO `product_sorting` VALUES
INSERT INTO `h1webblog_blog_translation` VALUES
INSERT INTO `neti_export_on_event` VALUES
INSERT INTO `state_machine_transition` VALUES
INSERT INTO `language` VALUES
INSERT INTO `country_translation` VALUES
INSERT INTO `product_sorting_translation` VALUES
INSERT INTO `product_search_config_field` VALUES
INSERT INTO `currency_translation` VALUES
INSERT INTO `number_range_type` VALUES
INSERT INTO `product_visibility` VALUES
INSERT INTO `customer_group_translation` VALUES
INSERT INTO `cms_section` VALUES
INSERT INTO `mail_header_footer_translation` VALUES
INSERT INTO `unit_translation` (`unit_id`, `language_id`, `short_code`, `name`, `custom_fields`, `created_at`, `updated_at`) VALUES
....
```

The problematic table is h1webblog_blog. In the old dump it was before that sudden switch of the complete-insert mode out of nowhere so I was able eto import it. In the new dump, the tables are dumped in a different order and it is exported after this sudden switch and causes problems.


I am also wondering how the table order in the dump is determined - does that have something with the foreign key constellations?

Because pretty much the only thing that changed is that 3 new tables were created. This could of course influcence the whole foreign key tree and maybe cause such a strange error to surface?

**Comment by amenk** [comment]
Must be this: As soon as a virtual column is found, it switches the complete inserts on:

https://github.com/ifsnop/mysqldump-php/blob/b232c2b6b2e250d716f936b0d11eda6bc2ec3df4/src/Ifsnop/Mysqldump/Mysqldump.php#L1274

Now that's really some side-effect of a getter :-)

**Comment by amenk** [comment]
Turns out this happend only in some rare cases with virtual + blob columns and the on the next table in the dump :-)

@guvra I created a pull request at the druidfi fork and hope it will be merged :-)

As soon as it is merged, shall we switch to the druidfi fork? Maybe you can already have a look at https://github.com/iMi-digital/gdpr-dump/commit/9916990e60c11f46cd2f338ebb3aeba498eaf796 -- of course the custom repository in composer.json will then not be necessary anymore.

**Comment by amenk** [comment]
Dev build in case anyone else want's to test.
[gdpr-dump-dev-fix78.zip](https://github.com/Smile-SA/gdpr-dump/files/10013615/gdpr-dump-dev-fix78.zip)

**Comment by amenk** [comment]
@guvra my fix was merged in the fork, shall we switch to that one?

**Comment by guvra** [comment]
@amenk Latest release is from Aug 24 (1.0.1).

I'd be willing to switch if they make a new release with the latest changes.

**Comment by guvra** [comment]
Solved by #79
